README.md
# RNAmodR.AlkAnilineSeq [![Build Status](https://travis-ci.com/FelixErnst/RNAmodR.AlkAnilineSeq.svg?branch=master)](https://travis-ci.com/FelixErnst/RNAmodR.AlkAnilineSeq) [![codecov](https://codecov.io/gh/FelixErnst/RNAmodR.AlkAnilineSeq/branch/master/graph/badge.svg)](https://codecov.io/gh/FelixErnst/RNAmodR.AlkAnilineSeq)

`RNAmodR.AlkAnilineSeq` uses classes and workflows from `RNAmodR` to detect 
`m7G`, `m3C` and `D` in RNASeq data specifically treated according to the 
AlkAnilineSeq protocol.

The vignette from the `RNAmodR.AlkAnilineSeq` package provides details on the
specific analysis of AlkAnilineSeq data. Have a look at the vignette from the
`RNAmodR` package for a general description on the provided workflows.

# Installation

The current version of the RNAmodR.AlkAnilineSeq package is available from
Bioconductor.

```
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("RNAmodR.AlkAnilineSeq")
library(RNAmodR.AlkAnilineSeq)
```

# Literature

- Marchand, Virginie, Lilia Ayadi, __Felix G. M. Ernst__, Jasmin Hertler,
Valérie Bourguignon-Igel, Adeline Galvanin, Annika Kotter, Mark Helm, 
__Denis L. J. Lafontaine__, and Yuri Motorin. 2018. “AlkAniline-Seq: Profiling 
of m7G and m3C Rna Modifications at Single Nucleotide Resolution.” Angewandte 
Chemie International Edition 57 (51): 16785–90. 
https://doi.org/10.1002/anie.201810946.


NEWS
Changes in version 0.99.0 (2019-04-29)
+ Submitted to Bioconductor

Changes in version 1.5.1 (2021-01-12)
+ bugfix for settings function


tests/testthat.R
library(testthat)
library(RNAmodR.AlkAnilineSeq)

test_check("RNAmodR.AlkAnilineSeq")


tests/testthat/test-AlkAnilineSeq.R
context("AlkAnilineSeq")
# .get_heatmap_data
test_that("AlkAnilineSeq:",{
  # argument normalization
  input <- list()
  actual <- RNAmodR.AlkAnilineSeq:::.norm_aas_args(input)
  expect_type(actual,"list")
  expect_named(actual,c("minCoverage",
                        "minReplicate",
                        "find.mod",
                        "minLength",
                        "minSignal",
                        "minScoreNC",
                        "minScoreSR",
                        "minScoreBaseScore",
                        "scoreOperator"))
  expect_error(RNAmodR.AlkAnilineSeq:::.norm_aas_args(list(minCoverage = 10)))
  expect_equal(RNAmodR.AlkAnilineSeq:::.norm_aas_args(list(minCoverage = 10L)),
               actual)
  expect_error(RNAmodR.AlkAnilineSeq:::.norm_aas_args(list(minReplicate = 1)))
  expect_equal(RNAmodR.AlkAnilineSeq:::.norm_aas_args(list(minReplicate = 1L)),
               actual)
  expect_error(RNAmodR.AlkAnilineSeq:::.norm_aas_args(list(minSignal = 10)))
  expect_equal(RNAmodR.AlkAnilineSeq:::.norm_aas_args(list(minSignal = 10L)),
               actual)
  expect_error(RNAmodR.AlkAnilineSeq:::.norm_aas_args(list(minLength = 9)))
  expect_equal(RNAmodR.AlkAnilineSeq:::.norm_aas_args(list(minLength = 9L)),
               actual)
  expect_error(RNAmodR.AlkAnilineSeq:::.norm_aas_args(list(minSignal = 10)))
  expect_equal(RNAmodR.AlkAnilineSeq:::.norm_aas_args(list(minSignal = 10L)),
               actual)
  expect_error(RNAmodR.AlkAnilineSeq:::.norm_aas_args(list(minScoreNC = "a")))
  expect_equal(RNAmodR.AlkAnilineSeq:::.norm_aas_args(list(minScoreNC = 50)),
               actual)
  expect_error(RNAmodR.AlkAnilineSeq:::.norm_aas_args(list(minScoreSR = 2L)))
  expect_equal(RNAmodR.AlkAnilineSeq:::.norm_aas_args(list(minScoreSR = 0.5)),
               actual)
  expect_error(RNAmodR.AlkAnilineSeq:::.norm_aas_args(list(scoreOperator = 2L)))
  expect_equal(RNAmodR.AlkAnilineSeq:::.norm_aas_args(list(scoreOperator = "&")),
               actual)
  data <- data.frame(scoreNC = "1", scoreSR = "a",
                     stringsAsFactors = FALSE)
  actual <- RNAmodR.AlkAnilineSeq:::.get_aas_scores(data)
  expect_type(actual,"list")
  expect_named(actual,c("score","scoreSR"))
  # settings
  data("msaas", package = "RNAmodR.AlkAnilineSeq")
  expect_type(settings(msaas),"list")
  expect_type(settings(msaas[[1]]),"list")
  expect_error(settings(msaas[[1]]) <- 1,
               "'value' has to be a named.")
  expect_error(settings(msaas[[1]]) <- c(1),
               "'value' has to be a named.")
  # settings(msaas[[1]]) <- c(minCoverage = 11L)
  # expect_equal(settings(msaas[[1]])$minCoverage, 11L)
  # apply_Np1_offset 
  mod <- aggregate(sequenceData(msaas[[1]]), condition = "Treated")$NormEnd5SequenceData
  actual <- RNAmodR.AlkAnilineSeq:::.apply_Np1_offset(mod)
  expect_named(actual)
  expect_equal(mod[[1]][5,1], 13L)
  expect_equal(actual[[1]][5,1], 20L)
  # .aggregate_aas 
  actual <- RNAmodR.AlkAnilineSeq:::.aggregate_aas(msaas[[1]]) 
  expect_named(actual)
  expect_s4_class(actual,"CompressedSplitDFrameList")
  expect_equivalent(actual[[1]][1,3], 0.1046218)
  expect_equal(colnames(actual[[1]]),c("ends","scoreNC","scoreSR","baseScore"))
  # findMod
  actual <- findMod(msaas[[1]])
  expect_s4_class(actual,"GRanges")
  expect_equal(colnames(mcols(actual)),c("mod","source","type","score",
                                         "scoreSR","Parent"))
  expect_equal(unique(mcols(actual)$mod),c("m7G","D","m3C"))
  expect_equal(length(actual),9L)
  # show
  # expect_output(expect_warning(show(msaas),
                               # "Settings were changed after data aggregation or"))
  # expect_warning(modifications(msaas),
                 # "Settings were changed after data aggregation or")
  expect_true(all(modifications(msaas)[[1]] == actual))
  # plotting
  expect_error(RNAmodR.AlkAnilineSeq:::.norm_viz_mod_aas_args(list(), "1"),
               "Type '1' is not valid")
  actual <- RNAmodR.AlkAnilineSeq:::.norm_viz_mod_aas_args(list(), "scoreNC")
  expect_type(actual,"list")
  expect_named(actual,c("type","colour"))
  expect_equal(actual[["type"]],"scoreNC")
  actual <- RNAmodR.AlkAnilineSeq:::.norm_viz_mod_aas_args(list(), "scoreSR")
  expect_equal(actual[["type"]],"scoreSR")
  actual <- RNAmodR.AlkAnilineSeq:::.norm_viz_mod_aas_args(list(), "ends")
  expect_equal(actual[["type"]],"ends")
  # getDataTrack
  expect_error(getDataTrack(msaas[[1]]),
               'argument "type" is missing, with no default')
  expect_error(getDataTrack(msaas[[1]],type="score"),
               "Type 'score' is not valid")
  expect_error(getDataTrack(msaas[[1]],type="scoreNC"),
               'argument "name" is missing')
  actual <- getDataTrack(msaas[[1]],name="1",type="scoreNC")
  expect_type(actual,"list")
  expect_named(actual,"scoreNC")
  expect_s4_class(actual[[1]],"DataTrack")
  # plotData
  actual <- plotData(msaas[[1]],name="1",type="scoreNC")
  expect_type(actual,"list")
  expect_named(actual,c("Normalized Cleavage","SequenceTrack","titles"))
  expect_s4_class(actual[[1]],"DataTrack")
  expect_s4_class(actual[[2]],"SequenceRNAStringSetTrack")
  expect_s4_class(actual[[3]],"ImageMap")
  actual <- plotData(msaas,name="1",type="scoreNC")
  expect_type(actual,"list")
  expect_s4_class(actual[[1]],"DataTrack")
  expect_s4_class(actual[[2]],"DataTrack")
  expect_s4_class(actual[[3]],"DataTrack")
  expect_s4_class(actual[[4]],"SequenceRNAStringSetTrack")
  expect_s4_class(actual[[5]],"ImageMap")
})


tests/testthat/test-AlkAnilineSeq-scores.R
context("AlkAnilineSeq base score")
# .get_heatmap_data
test_that("AlkAnilineSeq base score",{
  # base score selection
  df <- S4Vectors::DataFrame(means.treated.G = c("100","0"),
                             means.treated.A = c("0","100"),
                             means.treated.T = c("0","100"),
                             means.treated.C = c("0","100"))
  x <- IRanges::SplitDataFrameList(df,df)
  seq <- Biostrings::RNAStringSet(c("AG","CU"))
  actual <- RNAmodR.AlkAnilineSeq:::.selected_base_score(x,seq)
  expect_equal(actual,c(0,0,0,100))
})


R/RNAmodR.AlkAnilineSeq.R
#' @title RNAmodR.AlkAnilineSeq
#' 
#' @author Felix G M Ernst [aut], Denis L J Lafontaine [fnd]
#' 
#' @description
#' `RNAmodR.AlkAnilineSeq` implements the detection of 7-methyl guanosine,
#' 3-methyl cytidine and dihydrouridine from AlkAnilineseq data using the 
#' workflow and class the package `RNAmodR` provides.
#' 
#' @seealso Further details are described in the man pages of the 
#' \code{\link[RNAmodR:Modifier-class]{Modifier}} object and the vignettes.
#'
#' @docType package
#' @name RNAmodR.AlkAnilineSeq
NULL

#' @import methods
#' @import RNAmodR
#' @import S4Vectors
#' @import BiocGenerics
#' @import IRanges
#' @import GenomicRanges
#' @import Gviz
NULL

#' @name RNAmodR.AlkAnilineSeq-datasets
#' @title Example data in the RNAmodR.AlkAnilineSeq package
#' @description 
#' This contains an example ModifierSet object of type ModSetAlkAnilineSeq
#' @docType data
#' @format a \code{ModSetAlkAnilineSeq} instance
#' @keywords datasets
#' 
#' @usage data(msaas)
"msaas"

R/Modifier-AlkAnilineSeq-viz.R
#' @include RNAmodR.AlkAnilineSeq.R
#' @include Modifier-AlkAnilineSeq-class.R
NULL

RNAMODR_AAS_PLOT_DATA <- c("ends",
                           "scoreNC",
                           "scoreSR")
RNAMODR_AAS_PLOT_DATA_DEFAULT <- c("scoreNC")
RNAMODR_AAS_PLOT_DATA_NAMES <- c(ends = "5'-ends",
                                 scoreNC = "Normalized Cleavage",
                                 scoreSR = "Stop ratio")
RNAMODR_AAS_PLOT_DATA_COLOURS <- c(ends = "#FBB4AE",
                                   scoreNC = "#DECBE4",
                                   scoreSR = "#CCEBC5")

.norm_viz_mod_aas_args <- function(input, type){
  if(!all(type %in% RNAMODR_AAS_PLOT_DATA)){
    stop("Type '",type,"' is not valid. Valid types are: '",
         paste0(RNAMODR_AAS_PLOT_DATA, collapse = "','"),"'.",
         call. = FALSE)
  }
  colour <- input[["colour"]]
  if(!is.null(input[["colour"]])){
    colour <- RNAmodR:::.norm_viz_colour(input[["colour"]], type)
  } else {
    colour <- RNAMODR_AAS_PLOT_DATA_COLOURS[type]
  }
  input <- list(type = type,
                colour = colour)
  input
}

#' @rdname ModAlkAnilineSeq-functions
#' @export
setMethod(
  f = "getDataTrack",
  signature = signature(x = "ModAlkAnilineSeq"),
  definition = function(x, name, type, ...) {
    args <- .norm_viz_mod_aas_args(list(...), type)
    data <- RNAmodR:::.get_data_for_visualization(x, name)
    data <- unlist(data)
    dts <- lapply(
      args[["type"]],
      function(t){
        dt <- Gviz::DataTrack(range = data[,t],
                              groups = factor(t),
                              name = RNAMODR_AAS_PLOT_DATA_NAMES[t],
                              col = args[["colour"]][t],
                              type = "histogram")
        Gviz::displayPars(dt)$background.title <- "#FFFFFF"
        Gviz::displayPars(dt)$fontcolor.title <- "#000000"
        Gviz::displayPars(dt)$col.axis <- "#000000"
        Gviz::displayPars(dt) <- args[names(args) != "type"]
        dt
      })
    names(dts) <- args[["type"]]
    dts
  }
)

#' @rdname ModAlkAnilineSeq-functions
#' @export
setMethod(
  f = "plotDataByCoord",
  signature = signature(x = "ModAlkAnilineSeq",
                        coord = "GRanges"),
  definition = function(x, coord, type = c("ends","scoreNC","scoreSR"),
                        window.size = 15L, ...) {
    if(missing(type)){
      type <- RNAMODR_AAS_PLOT_DATA_DEFAULT
    }
    type <- match.arg(type,RNAMODR_AAS_PLOT_DATA, several.ok = TRUE)
    callNextMethod(x = x, coord = coord, type = type, window.size = window.size,
                   ...)
  }
)
#' @rdname ModAlkAnilineSeq-functions
#' @export
setMethod(
  f = "plotData",
  signature = signature(x = "ModAlkAnilineSeq"),
  definition = function(x, name, from, to, type = c("ends","scoreNC","scoreSR"),
                        ...) {
    if(missing(type)){
      type <- RNAMODR_AAS_PLOT_DATA_DEFAULT
    }
    type <- match.arg(type,RNAMODR_AAS_PLOT_DATA, several.ok = TRUE)
    callNextMethod(x = x, name = name, from = from, to = to, type = type, ...)
  }
)

#' @rdname ModAlkAnilineSeq-functions
#' @export
setMethod(
  f = "plotDataByCoord",
  signature = signature(x = "ModSetAlkAnilineSeq",
                        coord = "GRanges"),
  definition = function(x, coord, type = c("scoreNC","scoreSR","ends"),
                        window.size = 15L, ...) {
    if(missing(type)){
      type <- RNAMODR_AAS_PLOT_DATA_DEFAULT
    }
    type <- match.arg(type,RNAMODR_AAS_PLOT_DATA, several.ok = TRUE)
    callNextMethod(x = x, coord = coord, type = type, window.size = window.size,
                   ...)
  }
)
#' @rdname ModAlkAnilineSeq-functions
#' @export
setMethod(
  f = "plotData",
  signature = signature(x = "ModSetAlkAnilineSeq"),
  definition = function(x, name, from, to, type = c("scoreNC","scoreSR","ends"),
                        ...) {
    if(missing(type)){
      type <- RNAMODR_AAS_PLOT_DATA_DEFAULT
    }
    type <- match.arg(type,RNAMODR_AAS_PLOT_DATA, several.ok = TRUE)
    callNextMethod(x = x, name = name, from = from, to = to, type = type, ...)
  }
)


R/Modifier-AlkAnilineSeq-class.R
#' @include RNAmodR.AlkAnilineSeq.R
NULL

#' @name ModAlkAnilineSeq
#' @aliases AlkAnilineSeq ModAlkAnilineSeq ModSetAlkAnilineSeq
#' 
#' @author Felix G.M. Ernst [aut]
#' 
#' @title ModAlkAnilineSeq class to analyze AlkAnilineSeq data
#' 
#' @description 
#' 7-methyl guanosine (m7G), 3-methyl cytidine (m3C) and Dihydrouridine (D) 
#' are commonly found in rRNA and tRNA and can be detected classically by
#' primer extension analysis. However, since the modifications do not interfere
#' with Watson-Crick base pairing, a specific chemical treatment is employed
#' to cause strand breaks specifically at the modified positions.
#' 
#' This classical protocol was converted to a high throughput sequencing 
#' method call AlkAnilineSeq and allows modified position be detected by an
#' accumulation of 5'-ends at the N+1 position. Since the identify of the 
#' unmodified nucleotide is different for the three modified nucleotides, they
#' modification can be detected at the same time from the same samples.
#' 
#' \code{dataType} is \code{c("NormEnd3SequenceData","PileupSequenceData")}:
#' 
#' The \code{ModAlkAnilineSeq} class uses the  
#' \code{\link[RNAmodR:NormEndSequenceData-class]{NormEnd5SequenceData}}
#' class to store and aggregate data along the transcripts. This includes 
#' normalized values against the whole transcript (normalzed cleavage) and 
#' normalized values against the overlapping reads (stop ratio), which are used
#' to score for modified positions.
#' 
#' In addition the 
#' \code{\link[RNAmodR:PileupSequenceData-class]{PileupSequenceData}} class is
#' used as well, to check, whether the base is can be called according to the
#' expected sequence identity.
#' 
#' Only samples named \code{treated} are used for this analysis. Normalization 
#' to untreated samples is currently not used.
#' 
#' @param x the input which can be of the different types depending on whether
#' a \code{ModRiboMethSeq} or a \code{ModSetRiboMethSeq} object is to be 
#' constructed. For more information have a look at the documentation of
#' the \code{\link[RNAmodR:Modifier-class]{Modifier}} and 
#' \code{\link[RNAmodR:ModifierSet-class]{ModifierSet}} classes.
#' @param annotation annotation data, which must match the information contained
#' in the BAM files. This is parameter is only required if \code{x} if not a 
#' \code{Modifier} object.
#' @param sequences sequences matching the target sequences the reads were 
#' mapped onto. This must match the information contained in the BAM files. This
#' is parameter is only required if \code{x} if not a \code{Modifier} object.
#' @param seqinfo An optional \code{\link[GenomeInfoDb:Seqinfo-class]{Seqinfo}} 
#' argument or character vector, which can be coerced to one, to subset the 
#' sequences to be analyzed on a per chromosome basis.
#' @param ... Optional arguments overwriting default values, which are
#' \itemize{
#' \item{minLength:} {The minimal read length to be used for the analysis 
#' (default: \code{minLength = 9L}).}
#' \item{minSignal:} {The minimal signal at the position as integer value 
#' (default: \code{minSignal = 10L}). If the reaction is very specific a lower
#' value may need to be used}
#' \item{minScoreNC:} {minimum for score (normalized cleavage) to identify m7G,
#' m3C and D positions de novo (default: \code{minScoreNC = 50L})}
#' \item{minScoreSR:} {minimum for score (stop ration) to identify m7G, m3C and D
#' positions de novo (default: \code{minScoreSR = 0.5})}
#' \item{minScoreBaseScore:} {minimum score for base calling (0.0-1.0)  
#' (default: \code{minScoreSR = 0.9})}
#' \item{scoreOperator:} {how the minimal score should be used as logical 
#' operator. "&" requires all minimal values to be exceeded, whereas "|" detects
#' positions, if at least one minimal values is exceeded (default: 
#' \code{scoreOperator = "&"}).}
#' \item{other arguments} {which are passed on to 
#' \code{\link[RNAmodR:EndSequenceData-class]{End5SequenceData}}}
#' }
#' 
#' @return a \code{ModAlkAnilineSeq} or \code{ModSetAlkAnilineSeq} object
#' 
#' @references 
#' - Marchand V, Ayadi L, __Ernst FGM__, Hertler J, Bourguignon-Igel V,
#' Galvanin A, Kotter A, Helm M, __Lafontaine DLJ__, Motorin Y (2018): 
#' "AlkAniline-Seq: Profiling of m7 G and m3 C RNA Modifications at Single 
#' Nucleotide Resolution." Angewandte Chemie (International ed. in English) 57 
#' (51), P. 16785–16790. DOI: 
#' \href{https://doi.org/10.1002/anie.201810946}{10.1002/anie.201810946}.
#' 
#' @examples
#' library(RNAmodR.Data)
#' library(rtracklayer)
#' annotation <- GFF3File(RNAmodR.Data.example.AAS.gff3())
#' sequences <- RNAmodR.Data.example.AAS.fasta()
#' files <- list("wt" = c(treated = RNAmodR.Data.example.wt.1()),
#'               "Bud23del" = c(treated = RNAmodR.Data.example.bud23.1()),
#'               "Trm8del" = c(treated = RNAmodR.Data.example.trm8.1()))
#' # Creating a Modifier object of type ModRiboMethSeq
#' maas <- ModAlkAnilineSeq(files[[1]], annotation = annotation,
#'                          sequences = sequences)
#' # Creating a ModifierSet object of type ModSetRiboMethSeq
#' msaas <- ModSetAlkAnilineSeq(files, annotation = annotation,
#'                              sequences = sequences)
NULL

#' @rdname ModAlkAnilineSeq
#' @export
setClass("ModAlkAnilineSeq",
         contains = c("RNAModifier"),
         prototype = list(mod = c("m7G","m3C","D"),
                          score = "scoreNC",
                          dataType = c("NormEnd5SequenceData",
                                       "PileupSequenceData")))

# constructors -----------------------------------------------------------------

#' @rdname ModAlkAnilineSeq
#' @export
ModAlkAnilineSeq <- function(x, annotation = NA, sequences = NA, seqinfo = NA,
                             ...){
  RNAmodR::Modifier("ModAlkAnilineSeq", x = x, annotation = annotation, 
                    sequences = sequences, seqinfo = seqinfo, ...)
}

# settings ---------------------------------------------------------------------

#' @name ModAlkAnilineSeq-functions
#' @aliases aggregate modify settings plotData plotDataByCoord
#' 
#' @title Functions for ModAlkAnilineSeq
#' 
#' @description
#' All of the functions of \code{\link[RNAmodR:Modifier-class]{Modifier}} and
#' the \code{\link[RNAmodR:ModifierSet-class]{ModifierSet}} classes are
#' inherited by the \code{ModAlkAnilineSeq} and \code{ModSetAlkAnilineSeq}
#' classes.
#' 
#' @param x a \code{\link[RNAmodR:Modifier-class]{Modifier}} or a
#' \code{\link[RNAmodR:ModifierSet-class]{ModifierSet}} object. For more
#' details see also the man pages for the functions mentioned below.
#' @param value See \code{\link[RNAmodR:Modifier-class]{settings}}
#' @param coord,name,from,to,type,window.size,... See 
#' \code{\link[RNAmodR:plotData]{plotData}}.
#' 
#' @details 
#' \code{ModAlkAnilineSeq} specific arguments for \link{plotData}:
#' \itemize{
#' \item{\code{colour} - }{a named character vector of \code{length = 4} 
#' for the colours of the individual histograms. The names are expected to be 
#' \code{c("scoreNC","scoreSR")}}
#' }
#' 
#' @return 
#' \itemize{
#' \item{\code{settings}} {See 
#' \code{\link[RNAmodR:Modifier-functions]{settings}}.}
#' \item{\code{aggregate}} {See \code{\link[RNAmodR:aggregate]{aggregate}}.}
#' \item{\code{modify}} {See \code{\link[RNAmodR:modify]{modify}}.}
#' \item{\code{getDataTrack}} {a list of 
#' \code{\link[Gviz:DataTrack-class]{DataTrack}} object.}
#' \item{\code{plotData}} {See 
#' \code{\link[RNAmodR:plotData]{plotDataByCoord}}.}
#' \item{\code{plotDataByCoord}} {See 
#' \code{\link[RNAmodR:plotData]{plotDataByCoord}}.}
#' }
#' 
#' @importMethodsFrom RNAmodR modify aggregate settings plotData 
#' plotDataByCoord
#' 
#' @examples 
#' data(msaas,package="RNAmodR.AlkAnilineSeq")
#' maas <- msaas[[1]]
#' settings(maas)
#' aggregate(maas)
#' modify(maas)
#' getDataTrack(maas, "1", mainScore(maas))
NULL

.not_integer_bigger_than_zero <- RNAmodR:::.not_integer_bigger_than_zero
.not_numeric_bigger_zero <- RNAmodR:::.not_numeric_bigger_zero
.not_numeric_between_0_1 <- RNAmodR:::.not_numeric_between_0_1
.not_logical_operator <- RNAmodR:::.not_logical_operator

.ModAlkAnilineSeq_settings <- data.frame(
  variable = c("minLength",
               "minSignal",
               "minScoreNC",
               "minScoreSR",
               "minScoreBaseScore",
               "scoreOperator"),
  testFUN = c(".not_integer_bigger_than_zero",
              ".not_integer_bigger_than_zero",
              ".not_numeric_bigger_zero",
              ".not_numeric_between_0_1",
              ".not_numeric_between_0_1",
              ".not_logical_operator"),
  errorValue = c(TRUE,
                 TRUE,
                 TRUE,
                 TRUE,
                 TRUE,
                 TRUE),
  errorMessage = c("'minQuality' must be integer with a value higher than 0.",
                   "'minSignal' must be integer with a value higher than 0.",
                   "'minScoreNC' must be numeric with a value higher than 0.",
                   "'minScoreSR' must be numeric with a value between 0 and 1.",
                   "'minScoreBaseScore' must be numeric with a value between 0 and 1.",
                   "'scoreOperator' must be either '|' or '&'."),
  stringsAsFactors = FALSE)

.norm_aas_args <- function(input){
  minLength <- 9L # for all scores
  minSignal <- 10L # for all scores
  minScoreNC <- 50L # for score normalized cleavage
  minScoreSR <- 0.50 # for score stop ratio
  minScoreBaseScore <- 0.9 # for base scoring
  scoreOperator <- "&"
  args <- RNAmodR:::.norm_settings(input, .ModAlkAnilineSeq_settings, minLength,
                                   minSignal, minScoreNC, minScoreSR,
                                   minScoreBaseScore, scoreOperator)
  args <- c(RNAmodR:::.norm_Modifier_settings(input),
            args)
  args
}

#' @rdname ModAlkAnilineSeq-functions
#' @export
setReplaceMethod(f = "settings", 
    signature = signature(x = "ModAlkAnilineSeq"),
    definition = function(x, value){
        x <- callNextMethod()
        names <- names(value)
        value <- .norm_aas_args(value)
        x <- RNAmodR:::.add_settings_value(x, value, names)
        x
    }
)

# functions --------------------------------------------------------------------

.selected_base_score <- function(data,seq){
  seq <- IRanges::CharacterList(strsplit(as.character(seq),""))
  data <- data@unlistData
  # the order of baseSelect and baseColumns must be compatible
  # "conversion" T to U is performed with this
  baseSelect <- S4Vectors::DataFrame("G" = unlist(seq) == "G",
                                     "A" = unlist(seq) == "A",
                                     "U" = unlist(seq) == "U",
                                     "C" = unlist(seq) == "C")
  baseColumns <- which(colnames(data) %in% c("means.treated.G",
                                             "means.treated.A",
                                             "means.treated.T",
                                             "means.treated.C"))
  baseData <- as(data[,baseColumns],"NumericList")
  baseSelect <- as(baseSelect,"LogicalList")
  tmpBaseValues <- unname(baseData)[unname(baseSelect)]
  baseValues <- rep(0,nrow(data))
  baseValues[which(baseSelect[[1]])] <- tmpBaseValues[[1]]
  baseValues[which(baseSelect[[2]])] <- tmpBaseValues[[2]]
  baseValues[which(baseSelect[[3]])] <- tmpBaseValues[[3]]
  baseValues[which(baseSelect[[4]])] <- tmpBaseValues[[4]]
  baseValues[is.na(baseValues)] <- 0
  baseValues[is.infinite(baseValues)] <- 0
  baseValues
}

.apply_Np1_offset <- function(x){
  offsetAdd <- S4Vectors::DataFrame(as.list(rep(0,ncol(x@unlistData))))
  colnames(offsetAdd) <- colnames(unlist(x))
  x_rownames <- rownames(x)
  x <- x[IRanges::IntegerList(Map(seq.int,2L,lengths(x)))]
  x_offsetAdd <- IRanges::SplitDataFrameList(rep(list(offsetAdd),length(x)))
  x <- rbind(x,x_offsetAdd)
  rownames(x) <- x_rownames
  x
}

.aggregate_aas <- function(x){
  mod <- aggregate(sequenceData(x), condition = "Treated")
  if(is.null(names(mod))){
    names(mod) <- vapply(mod,class,character(1))
  }
  # get the NormEnd5SequenceData
  # shift each result one position upstream, since the signal has an N+1 offset
  endData <- .apply_Np1_offset(mod[["NormEnd5SequenceData"]])
  # get the PileupSequenceData
  # calculate base score
  pileupData <- mod[["PileupSequenceData"]]
  score <- .selected_base_score(pileupData,sequences(x))
  # subset to columns needed
  partitioning <- IRanges::PartitioningByEnd(endData)
  endData <- unlist(endData)[,c("means.treated.ends",
                             "means.treated.tx",
                             "means.treated.ol")]
  colnames(endData) <- c("ends","scoreNC","scoreSR")
  # add base score
  endData$baseScore <- score
  # split and return
  ans <- relist(endData, partitioning)
  rownames(ans) <- IRanges::CharacterList(RNAmodR:::.seqs_rl_strand(ranges(x)))
  ans
}

#' @rdname ModAlkAnilineSeq-functions
#' @export
setMethod(
  f = "aggregateData", 
  signature = signature(x = "ModAlkAnilineSeq"),
  definition = 
    function(x){
      .aggregate_aas(x)
    }
)

.get_aas_scores <- function(data){
  list(score = data$scoreNC,
       scoreSR = data$scoreSR)
}

.find_aas_modifications_in_data <- function(mod, letters, grl, signal,
                                            minScoreNC, minScoreSR,
                                            minScoreBaseScore,
                                            scoreOperator){
  modifications <- mapply(
    function(m,l,r,s){
      pos <- rownames(m)
      m <- m[!is.na(m$scoreNC) &
               !is.na(m$scoreSR) &
               !is.na(m$ends),,drop=FALSE]
      if(nrow(m) == 0L) return(NULL)
      m <- m[s,,drop=FALSE]
      if(nrow(m) == 0L) return(NULL)
      m <- m[mapply(Reduce,
                    rep(scoreOperator,nrow(m)),
                    m$scoreNC >= minScoreNC,
                    m$scoreSR >= minScoreSR,
                    m$baseScore >= minScoreBaseScore),,drop=FALSE]
      if(nrow(m) == 0L) return(NULL)
      l <- l[which(pos %in% rownames(m))]
      if(!any(l %in% c("G","C","U"))){
        return(NULL)
      }
      ansm7G <- RNAmodR:::constructModRanges(
        r,
        m[l == "G",],
        modType = "m7G",
        scoreFun = RNAmodR.AlkAnilineSeq:::.get_aas_scores,
        source = "RNAmodR.AlkAnilineSeq",
        type = "RNAMOD")
      ansD <- RNAmodR:::constructModRanges(
        r,
        m[l == "U",],
        modType = "D",
        scoreFun = RNAmodR.AlkAnilineSeq:::.get_aas_scores,
        source = "RNAmodR.AlkAnilineSeq",
        type = "RNAMOD")
      ansm3C <- RNAmodR:::constructModRanges(
        r,
        m[l == "C",],
        modType = "m3C",
        scoreFun = RNAmodR.AlkAnilineSeq:::.get_aas_scores,
        source = "RNAmodR.AlkAnilineSeq",
        type = "RNAMOD")
      ans <-  unlist(GenomicRanges::GRangesList(list(ansm7G,ansD,ansm3C)))
      ans <- ans[order(BiocGenerics::start(ans))]
      ans
    },
    mod,
    letters,
    grl,
    signal)
  f <- !vapply(modifications, is.null, logical(1))
  modifications <- mapply(
    function(m,name){
      m$Parent <- rep(name,length(m))
      m
    },
    modifications[f],
    names(grl)[f],
    SIMPLIFY = FALSE)
  modifications <- GenomicRanges::GRangesList(modifications)
  modifications
}

.find_aas_modifications <- function(x){
  if(!hasAggregateData(x)){
    stop("Something went wrong.")
  }
  grl <- ranges(x)
  letters <- IRanges::CharacterList(strsplit(as.character(sequences(x)),""))
  # get the aggregate data
  mod <- getAggregateData(x)
  # set up some arguments
  minSignal <- settings(x,"minSignal")
  minScoreNC <- settings(x,"minScoreNC")
  minScoreSR <- settings(x,"minScoreSR")
  minScoreBaseScore <- settings(x,"minScoreBaseScore")
  scoreOperator <- settings(x,"scoreOperator")
  # construct logical vector for passing the minSignal threshold
  signal <- mod[,"ends"] >= minSignal
  # find modifications
  modifications <- .find_aas_modifications_in_data(mod, letters, grl, signal,
                                                   minScoreNC, minScoreSR,
                                                   minScoreBaseScore,
                                                   scoreOperator)
  unname(unlist(modifications))
}

#' @rdname ModAlkAnilineSeq-functions
#' @export
setMethod("findMod",
          signature = c(x = "ModAlkAnilineSeq"),
          function(x){
            .find_aas_modifications(x)
          }
)

# ModSetAlkAnilineSeq ----------------------------------------------------------

#' @rdname ModAlkAnilineSeq
#' @export
setClass("ModSetAlkAnilineSeq",
         contains = "ModifierSet",
         prototype = list(elementType = "ModAlkAnilineSeq"))

#' @rdname ModAlkAnilineSeq
#' @export
ModSetAlkAnilineSeq <- function(x, annotation = NA, sequences = NA, 
                                seqinfo = NA, ...){
  RNAmodR::ModifierSet("ModAlkAnilineSeq", x, annotation = annotation,
                       sequences = sequences, seqinfo = seqinfo, ...)
}
